src/Http/Kernel.php

```diff
@@ -143,24 +155,28 @@ protected function defineLocale()
          */
         $pattern = '/^\/(' . implode('|', array_keys($locales['supported'])) . ')(\/|(?:$)|(?=\?))/';
 
-        $uri = array_get($_SERVER, 'REQUEST_URI', filter_input(INPUT_SERVER, 'REQUEST_URI', FILTER_SANITIZE_URL));
-
-        if (($hint === 'uri' || $hint === true) && preg_match($pattern, $uri, $matches)) {
-
-            $_SERVER['ORIGINAL_REQUEST_URI'] = $uri;
-            $_SERVER['REQUEST_URI']          = preg_replace($pattern, '/', $uri);
-
+        if (($hint === 'uri' || $hint === true) && preg_match($pattern, $originalRequestUri, $matches)) {
+            $request->server->set('REQUEST_URI', preg_replace($pattern, '/', $originalRequestUri));
+            $request->initialize(
+                $request->query->all(), 
+                $request->request->all(), 
+                $request->attributes->all(), 
+                $request->cookies->all(), 
+                $request->files->all(), 
+                $request->server->all(), 
+                $request->getContent()
+            );
             define('LOCALE', $matches[1]);
-
             return;
         }
+
     }
 
     /**
      * Rewrite the admin URI based on
      * configured admin URI segment.
      */
-    protected function rewriteAdmin()
+    protected function rewriteAdmin($request)
     {
         // Our admin segment.
         $segment = 'admin';
```

Comment:
Maybe you would replace this stuff to dynamic segment? If you have a bit of time, of course...

Reply:
Do you mean like `$segment = $request->segment(1);`